charm-slurmd/src/interface_slurmd.py

```diff
@@ -60,56 +63,58 @@ def __init__(self, charm, relation_name):
         )
 
     def _on_relation_created(self, event):
-        """Handle the relation-created event.
+        """
+        Handle the relation-created event.
 
-        Set the partition_name on the application relation data.
+        Set the node inventory and partition_name on the relation data.
         """
-        partition_name = self._charm.get_partition_name()
-        if not partition_name:
-            event.defer()
-            return
+        # Generate the inventory and set it on the relation data.
+        node_name = self._charm.hostname
+        node_addr = event.relation.data[self.model.unit]["ingress-address"]
+
+        inv = get_inventory(node_name, node_addr)
+        inv["new_node"] = True
+        self.node_inventory = inv
 
         if self.framework.model.unit.is_leader():
+            # Set the partition name on the application data.
+            partition_name = self._charm.get_partition_name()
+            if not partition_name:
+                event.defer()
+                return
+
             event.relation.data[self.model.app]["partition_name"] = \
                 partition_name
 
     def _on_relation_joined(self, event):
         """Handle the relation-joined event.
 
-        Get the munge_key from slurmctld and save it to the charm stored state.
+        Get the munge_key, slurmctld_host and slurmctld_port from slurmctld
+        and save it to the charm stored state.
         """
-        event_app_data = event.relation.data.get(event.app)
-        if not event_app_data:
-            event.defer()
-            return
+        app_data = event.relation.data[event.app]["munge_key"]
 
         # slurmctld sets the munge_key on the relation-created event
-        # which happens before relation-joined. We can almost guarantee that
-        # the munge key will exist at this point, but check for it just incase.
-        munge_key = event_app_data.get("munge_key")
-        if not munge_key:
-            event.defer()
-            return
-
-        # Store the munge_key in the charm's state
-        self._store_munge_key(munge_key)
-
-        # get slurmctld's hostname and port to enable configless
-        host = event_app_data.get("slurmctld_host")
-        port = event_app_data.get("slurmctld_port")
-        if not (host or port):
-            event.defer()
-        self._store_slurmctld_host_port(host, port)
-
+        # which happens before relation-joined. We can guarantee that
+        # the munge_key, slurmctld_host and slurmctld_port will exist
+        # at this point so retrieve them from the relation data and store
+        # them in the charm's stored state.
+        self._store_munge_key(app_data["munge_key"])
+        self._store_slurmctld_host_port(
+            app_data["slurmctld_host"], app_data["slurmctld_port"]
+        )
```

Comment:
What do you think about:

```suggestion
        self._store_slurmctld_host_port(app_data["slurmctld_host"], 
                                        app_data["slurmctld_port"])

```